packages/gatsby/src/db/loki/nodes.js

```diff
@@ -1,12 +1,354 @@
+  return nodeTypeColl.insert(node)
+}
+
+/**
+ * Updates a node in the DB. This works by removing all fields db
+ * fields ($loki, meta, and id) and then using `Object.assign` to copy
+ * all fields in `node` over `oldNode`. Therefore node should contain
+ * ALL fields and not just changes.
+ *
+ * If `oldNode` is not provided, it is assumed that `node` is already
+ * a loki node and we will simply call update on it
+ *
+ * @param {Object} node The new node information. This should be all
+ * the node information. Not just changes
+ * @param {Object} oldNode The old node to write over. Optional.
+ */
+function updateNode(node, oldNode) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let coll = getNodeTypeCollection(type)
+  if (!coll) {
+    invariant(coll, `${type} collection doesn't exist. When trying to update`)
+  }
+
+  if (oldNode) {
+    const lokiKeys = new Set([`$loki`, `meta`, `id`])
+    _.forEach(oldNode, (v, k) => {
+      if (!lokiKeys.has(k)) {
+        delete oldNode[k]
+      }
+    })
+    Object.assign(oldNode, node)
+    coll.update(oldNode)
+  } else {
+    coll.update(node)
+  }
+}
+
+/**
+ * Deletes a node from its type collection and removes its id ->
+ * collName mapping. Function is idempotent. If the node has already
+ * been deleted, this is a noop.
+ *
+ * @param {Object} the node to delete. Must have an `id` and
+ * `internal.type`
+ */
+function deleteNode(node) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let nodeTypeColl = getNodeTypeCollection(type)
+  if (!nodeTypeColl) {
+    invariant(
+      nodeTypeColl,
+      `${type} collection doesn't exist. When trying to delete`
+    )
+  }
+
+  if (nodeTypeColl.by(`id`, node.id)) {
+    const nodeMetaColl = getDb().getCollection(colls.nodeMeta.name)
+    invariant(nodeMetaColl, `Collection ${colls.nodeMeta.name} should exist`)
+    nodeMetaColl.findAndRemove({ id: node.id })
+    // TODO What if this `remove()` fails? We will have removed the id
+    // -> collName mapping, but not the actual node in the
+    // collection. Need to make this into a transaction
+    nodeTypeColl.remove(node)
+  }
+  // idempotent. Do nothing if node wasn't already in DB
+}
+
+/**
+ * deprecated
+ */
+function deleteNodes(nodes) {
+  for (const node of nodes) {
+    deleteNode(node)
+  }
 }
 
+/////////////////////////////////////////////////////////////////////
+// Reducer
+/////////////////////////////////////////////////////////////////////
+
+function reducer(state = new Map(), action) {
+  switch (action.type) {
+    case `DELETE_CACHE`:
+      deleteAll()
+      return new Map()
```

Comment:
does it make sense to create `new Map()`? we should be able to set state just to `null` and return same state (or `null`) again so garbage collector doesn't need to additional work that we won't be using?

Reply:
👍 fixed.